GPS: Extract Septentrio Driver Into Separate Module

flyingthingsintothings opened the pull request:
<!--

Thank you for your contribution!

Get early feedback through
- Dronecode Discord: https://discord.gg/dronecode
- PX4 Discuss: http://discuss.px4.io/
- opening a draft pr and sharing the link

-->

### Solved Problem
Making changes to the PX4 GPS drivers was hard because new features needed to be implemented for every driver. This isn't possible for most people as they don't have access to the hardware required for testing. During the previous PX4 community Q&A on Discord (2024-03-13), I discussed with @dagar what the options were to simplify this. The solution that was proposed was to separate all the drivers as they all want to specialize more than originally planned. This would allow the drivers to have their own parameters and implement features more rapidly. It would also simplify testing, as people who want to implement a new feature only need access to the hardware for a single driver.

Fixes \/

### Solution
- Extract the Septentrio GPS driver out of the submodule into its own directory, included in the PX4 autopilot repository.

### Changelog Entry
For release notes:
```
Feature/Bugfix XYZ
New parameter: XYZ_Z
Documentation: Need to clarify page ... / done, read docs.px4.io/...
```

### Alternatives
We could also ...

### Test coverage
- Unit/integration test: ...
- Simulation/hardware testing logs: https://review.px4.io/

### Manual Testing

- Pixhawk 6C & mosaic-go heading:
  - Automatic GNSS receiver configuration works
  - Position shows up in QGroundControl
  - Number of reported satellites is correct
  - Corrections from base station, forwarded by QGroundControl work
  - Automatic startup of driver works
  - Connection to receiver is stable and doesn't drop
  - `status` command works
  - Driver start and stop work

### Context
- [Forum post about developer call where this was discussed](https://discuss.px4.io/t/px4-community-q-a-march-13-2024/37190)
- [Original idea to split up driver](https://github.com/PX4/PX4-Autopilot/pull/22874#issuecomment-1992516897)
- [Teseo GPS driver used as reference](https://github.com/AlexKlimaj/PX4-Autopilot/tree/60a14ce240a09855e4ddf5e9690ca20a46e34aa2/src/drivers/teseo_gps)
- Other drivers in the PX4 autopilot repository were used as a guide (barometer, distance sensor...) because they are also split into separate drivers.

flyingthingsintothings commented:
This currently compiles and runs, but it has several bugs and is not ready. I'm just looking for overall feedback about the approach and structure while I'm resolving these bugs. If this isn't the correct approach I would love some guidance about what can be improved/changed. I also haven't committed the changes to remove usage of the Septentrio driver in the submodule as I've been focused on the extracted driver first.

Some questions:
- How would serial communication setup work when multiple submodules exist? Would there be a general GPS submodule responsible for forwarding calls to the actual driver modules? Is there another way to manage access to the serial ports?
- This uses code from the submodule for RTCM parsing. Should this be included in the PX4 autopilot project? Should it be in every driver that needs it? I don't see much code shared in the other drivers (barometer, distance sensor...), but I think sharing RTCM parsing makes sense as it's a standardized format.

dagar commented:
A few quick notes
 - move to srv/drivers/gnss to avoid confusion with drivers/gps
 - would a parameter prefix like GNSS_SBF work?

flyingthingsintothings commented on src/drivers/gnss/Kconfig:
Not sure what the best structure here is. I'm not very familiar with the Kconfig system so any help here is welcome.

flyingthingsintothings commented on src/drivers/gnss/septentrio/septentrio.cpp:
```diff
@@ -0,0 +1,1439 @@
+#define TIMEOUT_5HZ             500     ///< Timeout time in mS, 1000 mS (1Hz) + 300 mS delta for error
+#define RATE_MEASUREMENT_PERIOD 5_s     ///< Rate at which bandwith measurements are taken
+#define RECEIVER_BAUD_RATE      115200  ///< The baudrate of the serial connection to the receiver
+// TODO: This number seems wrong. It's also not clear why an ULL is created and casted to UL (time_t).
+#define GPS_EPOCH_SECS ((time_t)1234567890ULL)
```
I just copied this from `src/drivers/gps/devices/src/gps_helper.h` and changed the TODO a bit. If anyone has an idea what this does, I'd love to know 😄

flyingthingsintothings commented on src/drivers/gnss/septentrio/septentrio.cpp:
```diff
@@ -0,0 +1,1439 @@
+
+		decode_init();
+
+		// TODO: Not sure whether this is still correct when drivers are separate modules.
+		set_device_type(DRV_GPS_DEVTYPE_SBF);
```
It's not very clear to me what the `set_device_type()` method is for. Is it only for drivers with multiple backends or should it be called here as well? What does the value represent?

flyingthingsintothings commented on src/drivers/gnss/septentrio/septentrio.cpp:
```diff
@@ -0,0 +1,1439 @@
+		_last_timestamp_time = _report_gps_pos.timestamp;
+		_rate_count_vel++;
+		_rate_count_lat_lon++;
+		// NOTE: Isn't this just `ret |= (_msg_status == 7)`?
+		ret |= (_msg_status == 7) ? 1 : 0;
```
Just copied this, but it is written in a confusing way. I don't want to touch this code without asking as I'm not 100% certain how the parsing implementation works (the purpose of some variables isn't immediately clear and there is no documentation for them).

flyingthingsintothings commented on src/drivers/gnss/septentrio/septentrio.cpp:
```diff
@@ -0,0 +1,1439 @@
+	}
+
+	if (ret > 0) {
+		// NOTE: Isn't this always 0?
+		_report_gps_pos.timestamp_time_relative = static_cast<int32_t>(_last_timestamp_time - _report_gps_pos.timestamp);
```
Seems to always be 0 in the MAVLink inspector. I don't fully understand what the purpose of this line is as `_last_timestamp_time` is set to `_report_gps_pos.timestamp` a bit before this line, so it's always 0.

flyingthingsintothings commented on src/drivers/gnss/septentrio/septentrio.h:
```diff
@@ -0,0 +1,413 @@
+	uORB::SubscriptionMultiArray<gps_inject_data_s, gps_inject_data_s::MAX_INSTANCES> _orb_inject_data_sub{ORB_ID::gps_inject_data}; ///< uORB topic about data to inject to the receiver
+
+	// Reading and update rates
+	// NOTE: Both `_rate_vel` and `_rate_lat_lon` seem to be used in exactly the same way.
+	uint64_t _interval_rate_start{0};            ///< Start moment of measurement interval for position and velocity messages from the receiver
+	float    _rate{0.0f};                        ///< uORB position data publish rate (Hz)
+	uint8_t  _rate_count_vel;                    ///< Velocity messages from receiver in current measurement interval
+	uint8_t  _rate_count_lat_lon{};              ///< Position messages from receiver in current measurement interval
+	unsigned _last_rate_rtcm_injection_count{0}; ///< Counter for number of RTCM messages
+	float    _rate_vel{0.0f};                    ///< Velocity message rate from receiver in the last measurement interval (Hz)
+	float    _rate_lat_lon{0.0f};                ///< Position message rate from receiver in the last measurement interval (Hz)
```
These two values are always set in the exact same locations and seem to always contain the same value.

julianoes commented:
Sorry to be late to the party but how does this work alongside the GPS driver submodule? For instance, how does it work in terms of GPS parameters, auto detection, etc.?

flyingthingsintothings commented:
I used [the Teseo GPS module](https://github.com/AlexKlimaj/PX4-Autopilot/tree/60a14ce240a09855e4ddf5e9690ca20a46e34aa2/src/drivers/teseo_gps) by @AlexKlimaj and the other modules in PX4 as a guide while writing this. Most of the other modules seem to use addresses to share a single serial connection while talking to peripherals. The GPS module is different because it uses a 1:1 connection to the receiver. The parameters used to configure the receiver are all unique to the module. That way every module uses only the parameters it needs and it's more clear to users what parameters are actually used by the driver (unlike the current ones where the description mentions if a parameter is only used by a single driver). The parameter used to select the serial port needs to be configured by the user to decide what driver they want to use. If they enable the parameter in the GPS module, the GPS module will be started (would be the default). If they enable it in the Septentrio module, the Septentrio module will be started. If the same port is selected in both modules, a single driver is started. This is what I understood was meant to happen when I discussed this with @dagar. If this approach isn't correct, I can change it according to feedback.

For now the GPS module is not changed. It still contains the driver for Septentrio receivers. I can remove it in this pull request, or make a separate pull request to remove the code for Septentrio receivers from the GPS module (some would remain in the submodule because it's used by QGroundControl).

flyingthingsintothings commented:
I will update the documentation. I can even do that as a PR alongside this one if that is preferred, but I would like to finish the design and decisions for the new driver before I start rewriting the documentation. Otherwise any changes to the driver would mean rewriting work in the documentation.

For the duplicated code, I think the only realistic approach is to do it over time. I can already remove the Septentrio driver from the GPS module. The Septentrio module code can be removed using `makeconfig` so it does not take up extra space. If another approach is preferred, please let me know 🙂

Seungwon-Yeo commented:
Dooray! 메일 발송 실패 안내



  
  
    
    
  
  
    
메일 발송
실패 안내
    
  
  
    
        
          
***@***.***) 
          님께
보낸
메일이
전송되지
못하였습니다.
        
        
          실패 사유를 확인해보세요.
        
    
  
  
    
      
        
          
            * 받는 사람 : 
***@***.***) 
          
          
            * 발송 시간 : 
2024-03-28T19:46:08
          
          
            * 메일 제목 : 
Re: [PX4/PX4-Autopilot] GPS: Extract Septentrio Driver Into Separate Module (PR #22904)
          
          
            
              
                * 실패 사유 : 
              
              
                
대외비/기밀 메일은 외부 계정으로 발송이 제한되어 일부 메일 발송 실패(수신자가 전달 설정한 경우 포함)
                
              
            
                      
          
        
      
    
  
  
    
        
          이 메일은 발신전용으로 회신되지 않습니다.
          더 궁금하신 사항은
          ***@***.***
          으로 문의해 주시기 바랍니다.
        
    
  
  
    
        © Dooray!.

julianoes commented:
I'll leave the design review, etc. to @dagar but that sounds fine to me then.

Seungwon-Yeo commented:
Dooray! 메일 발송 실패 안내



  
  
    
    
  
  
    
메일 발송
실패 안내
    
  
  
    
        
          
***@***.***) 
          님께
보낸
메일이
전송되지
못하였습니다.
        
        
          실패 사유를 확인해보세요.
        
    
  
  
    
      
        
          
            * 받는 사람 : 
***@***.***) 
          
          
            * 발송 시간 : 
2024-03-29T05:20:39
          
          
            * 메일 제목 : 
Re: [PX4/PX4-Autopilot] GPS: Extract Septentrio Driver Into Separate Module (PR #22904)
          
          
            
              
                * 실패 사유 : 
              
              
                
대외비/기밀 메일은 외부 계정으로 발송이 제한되어 일부 메일 발송 실패(수신자가 전달 설정한 경우 포함)
                
              
            
                      
          
        
      
    
  
  
    
        
          이 메일은 발신전용으로 회신되지 않습니다.
          더 궁금하신 사항은
          ***@***.***
          으로 문의해 주시기 바랍니다.
        
    
  
  
    
        © Dooray!.

flyingthingsintothings commented:
Added back dual-receiver support which I originally left out when I extracted the driver. I also fixed the reset command, which is probably also not working in existing driver as it uses the same approach (the problem was the `SSSSSSSSSS` in front of the command without time in between).

AlexKlimaj commented:
Is this ready to test? I can test it on the ARK Mosaic GPS.

AlexKlimaj commented:
Got it running but it looks like its stuck in a loop? It also overrode my configuration.

```INFO  [septentrio] Serial port found: COM1
INFO  [septentrio] Automatic configuration finished
INFO  [septentrio] Serial port found: COM1
INFO  [septentrio] Automatic configuration finished
INFO  [septentrio] Serial port found: COM1
INFO  [septentrio] Automatic configuration finished
INFO  [septentrio] Serial port found: COM1
INFO  [septentrio] Automatic configuration finished
INFO  [septentrio] Serial port found: COM1
INFO  [septentrio] Automatic configuration finished
INFO  [septentrio] Serial port found: COM1
INFO  [septentrio] Automatic configuration finished
INFO  [septentrio] Serial port found: COM1
INFO  [septentrio] Automatic configuration finished
INFO  [septentrio] Serial port found: COM1
INFO  [septentrio] Automatic configuration finished
```

flyingthingsintothings commented:
@AlexKlimaj That's strange. I've been developing other features based on this branch so I've tested it for a few weeks now and it is stable for me. I copied the code from the original driver in the `GPS` module and applied minimal changes to reduce the chance of regressions. Maybe it's the change I applied to the new serial interface. I'll see whether I can reproduce this.

I'm not completely sure I understand what you mean when you say it overrode your configuration. Does it not take the configuration into account? Which configuration specifically?

AlexKlimaj commented:
On a fresh boot of the STM32 it looks like it works. However, it overwrites any custom configuration I have made. If you want to make it do automatic configuration, then anything a user might want to change need to be exposed in a parameter. You might also want to turn off automatic configuration by default.

```
bmp390 #0 on I2C bus 1 address 0x76
ERROR [SPI_I2C] iis2mdc: no instance started (no device on bus?)
bmm150 #0 on I2C bus 1 address 0x10
Starting GPS Port on /dev/ttyS0
WARN  [septentrio] GPS communication logging could not be initialized
INFO  [uavcannode] Node ID 125, bitrate 1000000

NuttShell (NSH) NuttX-11.0.0
nsh> WARN  [septentrio] No valid serial port detected
INFO  [septentrio] Serial port found: COM1
INFO  [septentrio] Serial port found: COM1
INFO  [septentrio] Automatic configuration finished

nsh>
nsh> septentrio status
INFO  [septentrio] Main GPS
INFO  [septentrio] status: OK, port: /dev/ttyS0
INFO  [septentrio] sat info: disabled
INFO  [septentrio] rate reading:                     0 B/s
INFO  [septentrio] rate position:                10.00 Hz
INFO  [septentrio] rate velocity:                10.00 Hz
INFO  [septentrio] rate publication:             10.00 Hz
INFO  [septentrio] rate RTCM injection:   0.00 Hz
 sensor_gps
    timestamp: 21029000 (0.031825 seconds ago)
    timestamp_sample: 0
    latitude_deg: -1145915590261.650000
    longitude_deg: -1145915590261.650000
    altitude_msl_m: 0.000000
    altitude_ellipsoid_m: -20000000000.000000
    time_utc_usec: 1301122103295000
    device_id: 10747904 (Type: 0xA4, UNKNOWN:0 (0x00))
    s_variance_m_s: -20000000000.00000
    c_variance_rad: 0.01745
    eph: 327.67499
    epv: 327.67499
    hdop: 0.00000
    vdop: 0.00000
    noise_per_ms: 0
    jamming_indicator: 0
    vel_m_s: 28284270592.00000
    vel_n_m_s: -20000000000.00000
    vel_e_m_s: -20000000000.00000
    vel_d_m_s: 20000000000.00000
    cog_rad: -349065856.00000
    timestamp_time_relative: 0
    heading: nan
    heading_offset: 0.00000
    heading_accuracy: 0.00000
    rtcm_injection_rate: 0.00000
    automatic_gain_control: 0
    fix_type: 0
    jamming_state: 0
    spoofing_state: 0
    vel_ned_valid: False
    satellites_used: 0
    selected_rtcm_instance: 0
    rtcm_crc_failed: False
    rtcm_msg_used: 0

nsh>
```

AlexKlimaj commented:
We also probably want auto bauding. To do high rate output that the module supports plus RTCM injection/output, I will want to use 921600 baud. Right now the driver is hard coded to 115200.

SeptentrioGNSS commented:
@AlexKlimaj , it is understandable that an auto configuration overwrites certain configurations (for now it is limited to Stream1 for the autopilot to receiver UART communication, Stream2 is saved for the adaptable logging and further it is things like receiverDynamics, antenna baseline if heading is available) [@flyingthingsintothings if I am missing CLIs here, feel free to mention it]
So could you explain if this goes beyond these parameters in your test case? 

Separate, having the default being auto config is just a matter of preference, some might turn it on for ease of use and others prefer it turned off by default to not overwrite anything unnecessarily but whatever the case, it is a matter of documentation and what it is set to, for now it is set on by default.

flyingthingsintothings commented:
@AlexKlimaj This new commit should add baud rate parameters back, as well as automatic baud rate detection. I tried to be as minimal as possible during the configuration.

I had implemented this before but removed it as it was causing some problems during the rest of the configuration. My goal was for the configuration to complete in a single call to `configure()` in a short amount of time (+- 1s). With this commit, it still completes in a single call to `configure()` for higher baud rates, but has more problems and takes longer at lower baud rates or different baud rates between the two receivers. I have spent quite a lot of time looking into this problem but I can't find what is going wrong. It's also not easy to debug so I think this is the best I can get it from my side. Maybe someone else sees what I'm doing wrong that causes problems, in which case I'd be happy to apply further patches.

I looked at the other drivers and they seem to do the same thing (except for the old SBF driver which doesn't have baud rate detection).

AlexKlimaj commented:
Ideally, the user sets the baud rate param to match what they have configured on the device. 

In your detect_receiver_baud_rate, you can speed it up by starting with the baudrate the user has specified.

Please make auto configuration of Stream1 an option or fully configurable via parameters. I would like to run it at 100Hz and this driver overwrites that config. 

Changing the baudrate through the web interface appears to cause the auto detection to fail.

```
INFO  [septentrio] `_uart` now running at 9600
INFO  [septentrio] `_uart` now running at 19200
INFO  [septentrio] `_uart` now running at 38400
INFO  [septentrio] `_uart` now running at 921600
WARN  [SerialImpl] Using non-standard baudrate: 1000000
INFO  [septentrio] `_uart` now running at 1000000
WARN  [SerialImpl] Using non-standard baudrate: 1500000
INFO  [septentrio] `_uart` now running at 1500000
INFO  [septentrio] CONFIG: failed baud detection
INFO  [septentrio] `_uart` now running at 115200
INFO  [septentrio] Detected baud rate: 115200
INFO  [septentrio] Serial port found: COM1
INFO  [septentrio] CONFIG: failed delete output
INFO  [septentrio] `_uart` now running at 115200
INFO  [septentrio] `_uart` now running at 230400
INFO  [septentrio] `_uart` now running at 57600
INFO  [septentrio] `_uart` now running at 460800
WARN  [SerialImpl] Using non-standard baudrate: 500000
INFO  [septentrio] `_uart` now running at 500000
WARN  [SerialImpl] Using non-standard baudrate: 576000
INFO  [septentrio] `_uart` now running at 576000
WARN  [SerialImpl] Using non-standard baudrate: 1200
INFO  [septentrio] `_uart` now running at 1200
WARN  [SerialImpl] Using non-standard baudrate: 2400
INFO  [septentrio] `_uart` now running at 2400
WARN  [SerialImpl] Using non-standard baudrate: 4800
INFO  [septentrio] `_uart` now running at 4800
```

flyingthingsintothings commented:
> Would this be easier to follow as a state machine?

I'm not sure. I will look a bit further into it and consider this as an option. For now I will try to solve the baud rate issue you mentioned and shared a log for.

> Please make auto configuration of Stream1 an option or fully configurable via parameters. I would like to run it at 100Hz and this driver overwrites that config.

Since this PR only extracts the existing driver into a new module, I didn't know whether it was OK to add extra functionality. There is already another PR that adds a parameter to the driver to disable automatic configuration, which can be merged after this one.

If it's better to immediately include that feature here (as it's not that big), I can do that as well and close the other PR.

AlexKlimaj commented:
Since this is a focused Septentrio driver, I would include everything in this PR to best represent and utilize the hardware. 

Take a look at the icm42688p driver for a good example of the state machine. Instead of running the main driver logic in a while loop, you would run it through a state machine like this. Its much easier to follow and debug.
https://github.com/PX4/PX4-Autopilot/blob/main/src/drivers/imu/invensense/icm42688p/ICM42688P.cpp#L151-L323

Also remember to run ```make format``` before you push to avoid the CI check_format failure.

flyingthingsintothings commented:
@AlexKlimaj I rewrote some parts entirely and copied other parts that worked fine. Hopefully this version works better and is more maintainable. I've been using this on different flight controllers and it works across all of them. The issue I had with the baud rates was because I changed baud rate right after sending a command, which sent the command at the new baud rate, causing configuration to fail. That should now be fixed.

There is one problem: The driver does not fit in the firmware for certain boards with the current board configurations. One example is the Pixhawk 6X, which overflows the available capacity by 0.30%. I don't know what to do about that, as I can't just disable other modules.

AlexKlimaj commented:
Can you make this run in the work queue? In its current form, it doesn't show up in any of the tools ie work_queue status or top.

AlexKlimaj commented on src/drivers/gnss/septentrio/septentrio.cpp:
```diff
@@ -0,0 +1,1702 @@
+
+	// We should definitely match baud rates and detect used port, but don't do other configuration if not requested.
+	// This will force input on the receiver. That shouldn't be a problem as it's on our own connection.
+	if (!_automatic_configuration) {
```
This needs to be after the baud rate detection.

flyingthingsintothings replied:
Do you mean right after line 742 with the `detect_receiver_baud_rate()`? Currently it's after baud rate detection, setting baud rate on flight controller and port detection as none of them touch the state of the receiver (only sends `getEchoMessage` and force input with `SSSSSSSSSS` which affects only the attached port). I've done it this way so there can be no confusion in the future about why sending commands in other parts of the driver doesn't work, as the baud rates wouldn't be matched. Automatic configuration doesn't mean that users couldn't send commands using the MAVLink console in a future use case for example. Also the current `septentrio reset <hot|warm|cold>` commands require matched baud rates already.

flyingthingsintothings commented:
I saw a question about why this wasn't turned into a work queue task instead. The question is already removed, but just for future reference: The driver needs to sleep at some points. I looked up whether suspending tasks on the work queue in a cooperative way is possible, but from [this forum post](https://discuss.px4.io/t/sleeping-in-the-work-queue/6304/4) that doesn't seem to be the case. The original, combined GPS drivers were also not scheduled on the work queue.

flyingthingsintothings commented:
The issue turned out to be when both the `GPS` module and `Septentrio` module were set to use the same physical port. This was something I had asked in one of the community Q&A calls during the very early development where I was told it wouldn't cause problems. If the `GPS` module isn't configured to use the same ports as the `Septentrio` one, everything works fine (tested on Pixhawk 6X and Cube Orange+ where issues occured).

flyingthingsintothings commented:
@mrpollo Could this PR get the label for the coordination call again to discuss blocking issues and progress?

flyingthingsintothings commented:
I already prepared the necessary documentation for this new driver. We are currently testing it to see what extra things need to be included to make it as user friendly as possible. I've documented all the parameters in there and cleaned the existing documentation about the hardware setup a bit. Once the documentation is finished I'll also make a pull request for that.

flyingthingsintothings commented:
@dagar Is there something else I should change about this PR or is it ready to be merged? I can make any changes if needed.

flyingthingsintothings commented:
I'm super excited to see this big change merged. Thank you to everyone who helped during the community Q&A sessions and with feedback here on this PR. I will continue to be available should any changes be required or when any bugs show up. I will also make sure to update the documentation very soon after I've made the final tweaks to the changes I have prepared for the user guide.
